File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> mp = new HashMap<>();
        mp.put('(',')');
        mp.put('{','}');
        mp.put('[',']');
        Stack<Character> st = new Stack<>();
        for(int i = 0;i<s.length();i++){
            char ch = s.charAt(i);
            if(ch == '('||ch == '{' || ch == '['){
                st.push(ch);
            }else{
                if(st.empty()) return false;
                char ch2 = st.pop();
                if(ch!=mp.get(ch2)) return false;
            }
        }
        return st.empty();
    }
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        valid = {
            '(' : ')',
            '[' : ']',
            '{' : '}'
        }

        for ch in s:
            if ch in valid:
                stack.append(ch)
            elif stack and valid[stack[-1]] == ch:
                stack.pop()
            else:
                return False
                # if stack:
                #     ch2 = stack.pop()
                #     if valid[ch] != stack[-1]:
                #         return False
                # else:
                #     return False

        return not stack


        